src/Mvc/Mvc.Core/src/ModelBinding/PrefixContainer.cs

```diff
@@ -201,32 +201,40 @@ private int BinarySearch(string prefix)
             {
                 Debug.Assert(candidate.StartsWith(prefix, StringComparison.OrdinalIgnoreCase));
 
-                // Okay, now we have a candidate that starts with the prefix. If the candidate is longer than
-                // the prefix, we need to look at the next character and see if it's a delimiter.
-                if (candidate.Length == prefix.Length)
-                {
-                    // Exact match
-                    return pivot;
-                }
-
-                var c = candidate[prefix.Length];
-                if (c == '.' || c == '[')
+                // At this point, we have identified a candidate that starts with the given prefix.
+                // If the candidate's length is greater than that of the prefix, we need to examine
+                // the character that immediately follows the prefix in the candidate string.
+                // This step is crucial to determine if the candidate is a valid prefix match.
+                // A valid prefix match occurs if the next character is either a delimiter ('.' or '['),
+                // indicating that the candidate is a sub-key of the prefix. If the next character
+                // is not a delimiter, it means the candidate contains additional characters that
+                // extend beyond the prefix without forming a valid hierarchical relationship,
+                // which should not qualify as a prefix match. Therefore, we will continue searching
+                // for valid matches in this case.
+
+                if (candidate.Length > prefix.Length)
```

Comment:
In order to safely access `candidate[prefix.Length]` the JIT emits a bound check. This bound check can be elided by 
```suggestion
                if ((uint)candidate.Length > (uint)prefix.Length)
```
See [sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHunGBLAHYYGAMQgQAFMXIAGBmGwCAJnyXYMMNE1kMADrABmfBFsHDdfAG4QMASi49O1Hi4Z8DDCQuWr1MAHQAMjACAOYYABYMAHx6hsZBIeER9s6u3E7p6WAR2FDyDAC88ooqahoA2vowRgiJYZEAugDcDlk8be1uHl5FxQDk/v0MhIQFhQMV/aldjp2z3MQA7HpWNq1pswC+8zw7m667iytw5Bsu+y6d9G5CDABCeVI63mV+WtJy1bWmdxbWdk6mSy7k8EgArmZbK9fBp6skYp5IUJbN8EsEGikjhxsS4cnlxiUfOUYFV4nUMckWriOgd2qDehMGINhqNxpNpjSMlzXMtVgDzgtLllhelsXzToLuPstkA==) for the difference in codegen.